pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,642 @@
+type ProfileSymbolizer struct {
+	logger  log.Logger
+	client  DebuginfodClient
+	store   DebugInfoStore
+	metrics *Metrics
+
+	symbolCache    *lru.Cache[string, []SymbolLocation]
+	debugInfoCache *ristretto.Cache
+}
+
+func NewFromConfig(_ context.Context, logger log.Logger, cfg Config, reg prometheus.Registerer, bucket objstore.Bucket) (*ProfileSymbolizer, error) {
+	metrics := NewMetrics(reg)
+
+	if bucket == nil {
+		return nil, fmt.Errorf("storage bucket is required for symbolizer")
+	}
+
+	prefixedBucket := objstore.NewPrefixedBucket(bucket, "symbolizer")
+
+	store := NewObjstoreDebugInfoStore(prefixedBucket, cfg.PersistentDebugInfoStore.MaxAge, metrics)
+
+	client, err := NewDebuginfodClient(logger, cfg.DebuginfodURL, metrics)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use configured cache sizes or defaults
+	symbolCacheSize := DefaultInMemorySymbolCacheSize
+	if cfg.InMemorySymbolCacheSize > 0 {
+		symbolCacheSize = cfg.InMemorySymbolCacheSize
+	}
+
+	debuginfoCacheSize := DefaultInMemoryDebuginfoCacheSize
+	if cfg.InMemoryDebuginfoCacheSize > 0 {
+		debuginfoCacheSize = cfg.InMemoryDebuginfoCacheSize
+	}
+
+	return NewProfileSymbolizer(logger, client, store, metrics, symbolCacheSize, debuginfoCacheSize)
+}
+
+func NewProfileSymbolizer(logger log.Logger, client DebuginfodClient, store DebugInfoStore, metrics *Metrics, symbolCacheSize int, debuginfoCacheSize int) (*ProfileSymbolizer, error) {
+	symbolCache, err := lru.New[string, []SymbolLocation](symbolCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create symbol cache: %w", err)
+	}
+
+	// Create Ristretto cache for debug info
+	debugInfoCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1e7,                       // number of keys to track frequency of (10M)
+		MaxCost:     int64(debuginfoCacheSize), // maximum cost of cache
+		BufferItems: 64,                        // number of keys per Get buffer
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create debug info cache: %w", err)
+	}
+
+	return &ProfileSymbolizer{
+		logger:         logger,
+		client:         client,
+		store:          store,
+		metrics:        metrics,
+		symbolCache:    symbolCache,
+		debugInfoCache: debugInfoCache,
+	}, nil
+}
+
+func (s *ProfileSymbolizer) SymbolizePprof(ctx context.Context, profile *googlev1.Profile) error {
+	start := time.Now()
+	status := StatusSuccess
+	defer func() {
+		s.metrics.profileSymbolization.WithLabelValues(status).Observe(time.Since(start).Seconds())
+	}()
+
+	if !s.NeedsSymbolization(profile) {
+		status = "already_symbolized"
+		level.Error(s.logger).Log("msg", "Symbolizer exited since profile don't need symbolization")
+		return nil
+	}
+
+	// Group locations by mapping ID
+	type locToSymbolize struct {
+		idx int
+		loc *googlev1.Location
+	}
+	locsByMapping := make(map[uint64][]locToSymbolize)
+
+	for i, loc := range profile.Location {
+		if loc.MappingId > 0 {
+			mapping := profile.Mapping[loc.MappingId-1]
```

Comment:
IMO this is incorrect. you can not use ID for index into Mapping/Function/Location tables

Reply:
If the pprof is generated by the backend, it should work. This optimization is assumed along the entire read path

Reply:
TIL